### DISCUSSION TITLE
How to access and pass as a function parameter nested json objects in for loop?

### DISCUSSION THREAD

**Discussion by dreambeyondorange** [nlohmann/json#2631] [Q&A] [answered]
I send a sequence of several messages in a json object like this:
```json
 {
    "msg1" : {
        "time_since_epoch_ms" : 1612969709407,
        "timestamp" : "Wed Feb 10 10:08:29 2021\n",
        "type" : 1305,
        "value" : -999.0
    }
    "msg2" : {
        "time_since_epoch_ms" : 1612969709407,
        "timestamp" : "Wed Feb 10 10:08:29 2021\n",
        "type" : 1305,
        "value" : -999.0
    }
} 
```
And I'd like to iterate over these messages and pass the nested json to a function.

I keep running into this error during linking undefined reference to `<function name>(nlohmann::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long, unsigned long, double, std::allocator, nlohmann::adl_serializer, std::vector<unsigned char, std::allocator<unsigned char> > >)

I have a function 
`return_t func(nlohmann::json msg);`

And I've tried using range based loops like this
`for (const auto &msg : incomingJSON->front().items())`
with msg.value() as my nested object but when i try to call the function it gives me the above error no matter how i try to call it
`func(msg.value());`
`func(json::parse(msg.value()));`

I also tried range based loops like below, which works fine for access within the loop to the objects, but I cannot pass the iterator object as json to another function.
```cpp
for (auto &msg : json_object) {
                switch ((int)msg["type"]) {
...
                      func(msg);
```

What is the right way to do this?

**Comment by nlohmann** [comment, marked as answer]
The return value of `items()` is not `nlohmann::json`. You would need to use `declytype` to derive the type for the definition of `func`.

**Reply by nlohmann** [reply]
@dreambeyondorange Do you need further assistance?

**Comment by dreambeyondorange** [comment]
No, I have found a solution ty!
